server/src/main/java/org/opensearch/index/shard/RemoteSegmentTrackerListener.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.shard;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.search.ReferenceManager;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
+import org.opensearch.common.concurrent.GatedCloseable;
+import org.opensearch.common.logging.Loggers;
+import org.opensearch.index.remote.RemoteRefreshSegmentTracker;
+import org.opensearch.index.store.RemoteSegmentStoreDirectory;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * This listener updates the remote segment tracker with the segment files of the most recent refresh. This is helpful in
+ * determining the lag and hence applying rejection on lagging remote uploads.
+ *
+ * @opensearch.internal
+ */
+public class RemoteSegmentTrackerListener implements ReferenceManager.RefreshListener {
+
+    private final Logger logger;
+    private final IndexShard indexShard;
+    private final RemoteRefreshSegmentTracker segmentTracker;
+    private final RemoteSegmentStoreDirectory remoteDirectory;
+    private final Directory storeDirectory;
+    private long primaryTerm;
+
+    public RemoteSegmentTrackerListener(IndexShard indexShard, RemoteRefreshSegmentTracker segmentTracker) {
+        this.indexShard = indexShard;
+        this.segmentTracker = segmentTracker;
+        logger = Loggers.getLogger(getClass(), indexShard.shardId());
+        storeDirectory = indexShard.store().directory();
+        remoteDirectory = (RemoteSegmentStoreDirectory) ((FilterDirectory) ((FilterDirectory) indexShard.remoteStore().directory())
+            .getDelegate()).getDelegate();
+        if (indexShard.routingEntry().primary()) {
+            try {
+                this.remoteDirectory.init();
+            } catch (IOException e) {
+                logger.error("Exception while initialising RemoteSegmentStoreDirectory", e);
+            }
+        }
+        this.primaryTerm = remoteDirectory.getPrimaryTermAtInit();
+    }
+
+    @Override
+    public void beforeRefresh() throws IOException {}
+
+    @Override
+    public void afterRefresh(boolean didRefresh) throws IOException {
+        if (didRefresh
```

Comment:
These are the same conditions that we check in `RemoteStoreRefreshListener` as well, right? Do they need to be same in both these listeners?

Reply:
> These are the same conditions that we check in RemoteStoreRefreshListener as well, right?

Yes

> Do they need to be same in both these listeners?

Yes, they need to be same to ensure that the local tracker is getting updated on same conditions as it was being done earlier.